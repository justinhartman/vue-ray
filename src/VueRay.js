const { Ray } = require('node-ray/web');

class VueRay extends Ray {
    ref(name) {
        if (this.component) {
            this.html(this.component.$refs[name].innerHTML);
        }
    }
}

export default VueRay;
