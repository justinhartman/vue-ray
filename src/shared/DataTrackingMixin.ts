export const DataTrackingMixin = {
    data() {
        return {
            trackingRays: {},
            trackingStops: {},
        };
    },
};

export default DataTrackingMixin;
