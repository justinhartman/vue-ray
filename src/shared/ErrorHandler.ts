const errorHandler = (err: any, vm: any) => {
    // pretty-print the error message and stack trace
    const stack = err.stack
        .replace(
            /^(\w+):(.+)$/m,
            '<div class="flex bg-white border-0">' +
                '<div class="flex-row bg-white text-red-400 pr-1 pl-1 border border-1 border-red-400">$1</div>' +
                '<div style="padding-top: 0.25em;" class="flex-row">$2</div>' +
                '</div>'
        )
        .replace(/\n/, 'Stack Trace:\n')
        .replace(/ at ([^ ]+) (.*)$/gm, '<div class="inline"> at <span class="text-blue-700">$1</span> $2</div>'); // eslint-disable-line

    vm.$ray().sendCustom(`<pre>${stack}</pre>`, 'Error').small().red();
};

export default errorHandler;
