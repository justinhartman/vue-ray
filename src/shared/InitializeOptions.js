const { Ray } = require('node-ray/web');
import errorHandler from './ErrorHandler';

export const initializeOptions = (options, vueConfig) => {
    if (options.interceptErrors === true) {
        vueConfig.errorHandler = errorHandler;
    }

    let host = 'localhost',
        port = 23517;

    if (typeof options['host'] === 'string') {
        host = options.host;
    }

    if (typeof options['port'] === 'number') {
        port = options.port;
    }

    Ray.useDefaultSettings({ host, port });
};

export default initializeOptions;
