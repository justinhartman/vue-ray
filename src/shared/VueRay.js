const { Ray } = require('node-ray/web');

class VueRay extends Ray {
    data() {
        if (this.component) {
            this.table([this.component.$data]);
        }
    }

    props() {
        if (this.component) {
            this.table([this.component.$props]);
        }
    }

    ref(name) {
        if (this.component) {
            this.html(this.component.$refs[name].innerHTML);
        }
    }
}

export default VueRay;
