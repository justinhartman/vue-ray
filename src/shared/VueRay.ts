/* eslint-disable no-unused-vars */

/// <reference path="../../node_modules/node-ray/dist/web.d.ts" />

const { Ray } = require('node-ray/web');
//import { Ray } from 'node-ray/web';

export class VueRay extends Ray {
    public static show_component_lifecycles: string[] = [];
    //public component: any = { $refs: {} };
    public watch: CallableFunction = () => {};

    // static create(client?: any | null, uuid?: string | null): this {
    //     return Ray.create(client, uuid);
    // }

    public data(): void {
        if (this.component) {
            let data = Object.assign({}, this.component.$data);

            // remove Vue Observers to avoid recursive issues
            delete data['trackingRays'];
            delete data['trackingStops'];

            data = Object.freeze(data);

            this.table([data], 'Data');
        }
    }

    public props(): void {
        if (this.component) {
            this.table([this.component.$props]);
        }
    }

    public ref(name: string): void {
        if (this.component) {
            this.html(this.component.$refs[name].innerHTML);
        }
    }

    public track(name: string): void {
        if (this.component && typeof this.component.trackingRays[name] === 'undefined') {
            this.component.trackingRays[name] = this.component.$ray(name);

            const onTrackedUpdate = (value: any, previousValue: any) => {
                this.component.trackingRays[name] = this.component.trackingRays[name].send({
                    info: {
                        variable: name,
                        tracking: true,
                    },
                    trackingData: {
                        value: value,
                        previous: previousValue,
                    },
                });
            };

            this.component.trackingStops[name] = this.watch(() => this.component[name], onTrackedUpdate, { deep: true });
        }
    }

    public untrack(name: string): void {
        if (this.component && typeof this.component.trackingStops[name] !== 'undefined') {
            this.component.trackingRays[name] = this.component.trackingRays[name].send({
                info: {
                    variable: name,
                    tracking: false,
                },
                trackedData: {
                    value: this.component[name],
                },
            });

            this.component.trackingStops[name]();

            delete this.component.trackingStops[name];
            delete this.component.trackingRays[name];
        }
    }

    public static showComponentLifecycles(names: string[]): void {
        if (Array.isArray(VueRay.show_component_lifecycles)) {
            VueRay.show_component_lifecycles.push(...names);
        }
    }

    public static stopShowingComponentLifecycles(names: string[] = []): void {
        if (!names.length) {
            VueRay.show_component_lifecycles = [];
            return;
        }

        VueRay.show_component_lifecycles = VueRay.show_component_lifecycles.filter(name => names.includes(name));
    }
}

export default VueRay;
