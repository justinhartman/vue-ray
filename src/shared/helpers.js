import { PackageInfo } from './PackageInfo';

export const createPackageMetaProperty = obj => {
    obj.$rayMeta = PackageInfo;
};
