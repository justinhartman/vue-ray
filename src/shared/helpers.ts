import { PackageInfo } from './PackageInfo';
import multimatch from 'multimatch';

// @ts-ignore
export const createPackageMetaProperty = obj => {
    obj.$rayMeta = PackageInfo;
};

export const encodeHtmlEntities = (str: string) => {
    const escapeChars: Record<string, string> = {
        '¢': 'cent',
        '£': 'pound',
        '¥': 'yen',
        '€': 'euro',
        '©': 'copy',
        '®': 'reg',
        '<': 'lt',
        '>': 'gt',
        '"': 'quot',
        '&': 'amp',
        "'": '#39',
    };

    const chars: string[] = Object.keys(escapeChars);
    const regex = new RegExp(`[${chars.join('')}]`, 'g');

    return str.replace(regex, m => `&${escapeChars[m]};`);
};

export const filterObjectByKeys = (obj: any, includeKeyPatterns: string[]) => {
    const result: any = {};

    Object.keys(obj).forEach(key => {
        multimatch(key, includeKeyPatterns).forEach(match => {
            result[match] = obj[match];
        });
    });

    return result;
};
