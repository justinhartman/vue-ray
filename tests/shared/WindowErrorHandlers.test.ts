/* eslint-disable no-undef */
/* eslint-disable no-unused-vars */

import { VueRay } from '../../src/shared/VueRay';
import { FakeClient } from '../TestClasses/FakeClient';

import {
    setRayInstance,
    onErrorHandler,
    onUnhandledRejectionEventHandler,
    installWindowErrorHandlers,
} from '../../src/shared/WindowErrorHandlers';

let ray: VueRay, client: FakeClient;

beforeEach(() => {
    client = new FakeClient(3000, 'localhost');
    // @ts-ignore
    ray = VueRay.create(client, 'fakeUuid');

    ray.component = {
        $data: { one: 1 },
        $props: { title: 'abc' },
        $refs: { test1: { innerHTML: '<em>123</em>' } },
        trackingRays: {},
        trackingStops: {},
        $ray(...args: any[]) {
            // @ts-ignore
            return ray.send(...args);
        },
    };

    setRayInstance(ray.component.$ray);

    ray.data();
});

it('installs error handlers onto the window object', () => {
    const win = {
        onerror: null,
        unhandledrejection: null,
    };

    installWindowErrorHandlers(win, ray);

    expect(win.onerror).not.toBeNull();
    expect(win.unhandledrejection).not.toBeNull();
});

it('sends window errors to ray', () => {
    const win = {
        onerror: null,
        unhandledrejection: null,
    };

    installWindowErrorHandlers(win);

    onErrorHandler('test', 'source test', 12, 2, {});

    expect(client.sentPayloads().length).toBeGreaterThan(0);
    expect(client.sentPayloads()).toMatchSnapshot();
});

it('sends unhandled rejection errors to ray', () => {
    const win = {
        onerror: null,
        unhandledrejection: null,
    };

    const event = {
        reason: 'test',
    };

    installWindowErrorHandlers(win);

    onUnhandledRejectionEventHandler(event);

    expect(client.sentPayloads().length).toBeGreaterThan(0);
    expect(client.sentPayloads()).toMatchSnapshot();
});
