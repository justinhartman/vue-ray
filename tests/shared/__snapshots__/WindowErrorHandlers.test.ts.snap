// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sends unhandled rejection errors to ray 1`] = `
Array [
  Object {
    "meta": Array [],
    "payloads": Array [
      Object {
        "content": Object {
          "label": "Data",
          "values": Array [
            "<code style=\\"font-size: 0.8rem!important;\\"><span style=\\"font-size: 0.8rem!important;\\" class=\\"text-yellow-600 bold\\">Object</span>&nbsp;<code style=\\"font-size: 0.8rem!important;\\" class=\\"text-gray-600 \\"><span style=\\"font-size: 0.8rem!important;\\" class=\\"text-orange-400 bold\\">{</span><br>&nbsp;&nbsp;&nbsp;&nbsp;<code style=\\"font-size: 0.8rem!important;\\" class=\\"bold text-green-600 p-0\\">\\"one\\"</code><span style=\\"font-size: 0.8rem!important;\\" class=\\"text-orange-400 bold\\">:&nbsp;</span>1<span style=\\"font-size: 0.8rem!important;\\" class=\\"text-orange-400 bold\\">,</span><br><span style=\\"font-size: 0.8rem!important;\\" class=\\"text-orange-400 bold\\">}</span></code></code>",
          ],
        },
        "origin": Object {
          "file": "web.cjs.js",
          "function_name": "xxxx",
          "hostname": "remote",
          "line_number": 999,
        },
        "type": "table",
      },
    ],
    "uuid": "fakeUuid",
  },
  Object {
    "meta": Array [],
    "payloads": Array [
      Object {
        "content": Object {
          "content": "<pre>
    <div class=\\"flex bg-white border-0\\">
        <div class=\\"flex-row bg-white text-red-400 pr-1 pl-1 border border-1 border-red-400\\">test</div>
        <div style=\\"padding-top: 0.25em;\\" class=\\"flex-row\\"><code>{\\"reason\\":\\"test\\"}</code></div>
    </div></pre>",
          "label": "HTML",
        },
        "origin": Object {
          "file": "web.cjs.js",
          "function_name": "xxxx",
          "hostname": "remote",
          "line_number": 999,
        },
        "type": "custom",
      },
    ],
    "uuid": "fakeUuid",
  },
  Object {
    "meta": Array [],
    "payloads": Array [
      Object {
        "content": Object {
          "size": "sm",
        },
        "origin": Object {
          "file": "web.cjs.js",
          "function_name": "xxxx",
          "hostname": "remote",
          "line_number": 999,
        },
        "type": "size",
      },
    ],
    "uuid": "fakeUuid",
  },
  Object {
    "meta": Array [],
    "payloads": Array [
      Object {
        "content": Object {
          "color": "red",
        },
        "origin": Object {
          "file": "web.cjs.js",
          "function_name": "xxxx",
          "hostname": "remote",
          "line_number": 999,
        },
        "type": "color",
      },
    ],
    "uuid": "fakeUuid",
  },
]
`;

exports[`sends window errors to ray 1`] = `
Array [
  Object {
    "meta": Array [],
    "payloads": Array [
      Object {
        "content": Object {
          "label": "Data",
          "values": Array [
            "<code style=\\"font-size: 0.8rem!important;\\"><span style=\\"font-size: 0.8rem!important;\\" class=\\"text-yellow-600 bold\\">Object</span>&nbsp;<code style=\\"font-size: 0.8rem!important;\\" class=\\"text-gray-600 \\"><span style=\\"font-size: 0.8rem!important;\\" class=\\"text-orange-400 bold\\">{</span><br>&nbsp;&nbsp;&nbsp;&nbsp;<code style=\\"font-size: 0.8rem!important;\\" class=\\"bold text-green-600 p-0\\">\\"one\\"</code><span style=\\"font-size: 0.8rem!important;\\" class=\\"text-orange-400 bold\\">:&nbsp;</span>1<span style=\\"font-size: 0.8rem!important;\\" class=\\"text-orange-400 bold\\">,</span><br><span style=\\"font-size: 0.8rem!important;\\" class=\\"text-orange-400 bold\\">}</span></code></code>",
          ],
        },
        "origin": Object {
          "file": "web.cjs.js",
          "function_name": "xxxx",
          "hostname": "remote",
          "line_number": 999,
        },
        "type": "table",
      },
    ],
    "uuid": "fakeUuid",
  },
]
`;
